### PR TITLE
(TK-187) Port ConfigDocument and tests

### PR DESCRIPTION
**Pull request by fpringvaldsen**
Port all the functionality around the `ConfigDocument` class along with all the relevant tests.

### CODE REVIEW

**Comment by fpringvaldsen** [comment]
This PR also contains changes from https://github.com/typesafehub/config/pull/304

File: lib/hocon/impl/config_document_parser.rb [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+              " to be part of the value for '#{previous_field_name}', " +
+              "try enclosing the value in double quotes"
+        else
+          part = "#{message} (if you intended #{bad_token}" +
+              " to be part of a key or string value, " +
+              "try enclosing the key or value in double quotes"
+        end
+      end
+
+      # Don't special case to throw a message about changing the file to .properties, since
```
Comment (jpinsonault) [inline lib/hocon/impl/config_document_parser.rb]: Is this supposed to be "Don't have a special case"?

File: spec/unit/typesafe/config/config_document_parser_spec.rb [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+         }'}
+      include_examples "parse test"
+    end
+
+    context "empty array" do
+      let (:orig_text) { "[]" }
+      include_examples "parse test"
+    end
+
+    context "empty array" do
```
Comment (jpinsonault) [inline spec/unit/typesafe/config/config_document_parser_spec.rb]: Not sure if this should be "empty array"

**Comment by jpinsonault** [comment]
Just a couple minor comments, otherwise looks good. 

Also thank you for all the helpful comments in the code. Helped a ton

**Pull request merged**